### ISSUE TITLE
dash in template vairable name causes Flask hang.

### ISSUE THREAD

**Issue by samuelchen** [pallets/flask#3516]
Hi,

I was facing an issue when using incorrect variable name in `Flask` template. An dash "-" linked variable name in template cause Flask to hang without throwing any exception.

At first I doubt it's `Jinja` issue but I found Jinja raised the exception and `Werkzeug` handled it.

Eventually, I found an endless loop in `Python\Lib\traceback.py` to walk trackback stacks. The 
trackbacks are stored in a link which nodes point to next. In `flask\templating.py`, 
`flask\app.py` and `werkzeug\debug\tbtools.py`, the exception was catched. In 
`handle_exception()` of `flask\app.py`, `Werkzeug` wants to log it and print it out. 

The trackback link has looped reference as below picture.

![](http://blog.samuelchen.net/Flask-Werkzeug-templating-caused-trackback-loop-issue/flask_tb_next_loop_ref.png)

### Expected Behavior
Render variable out or throw exception out.

```python
import sys
from flask import Flask, request, render_template_string, make_response, url_for, redirect

### {{ font-size }} causes app hang ###

app = Flask(__name__)

template_page = '''
<html>
<head>
<meta charset="{{ encoding }}" />
<meta http-equiv="content-type" content="text/html; charset={{ encoding }}" />
<style type="text/css">
  body, pre, p, div, input, h1,h2,h3,h4,h5 {
    font-family : Consolas, Courier New;
  }
</style>
</head>
<body>
    <div id="wrapper" style="font-size:{{ font-size }}em">
        {{ content | safe }}
    </div>
</body>'''


@app.route('/', methods=['GET'])
def index():
    context = {
        "encoding": 'utf-8',
        "font-size": 0.9,
        "content": 'hello flask'
    }
    resp = make_response(render_template_string(template_page, **context))
    resp.headers['Content-Type'] = 'text/html; charset=utf-8'
    return resp


if __name__ == "__main__":
    # app.run(debug=True, host='0.0.0.0', port=8080)

    app.run()
```

### Actual Behavior
App hangs.

```pytb
N/A
```

### Environment

* Python version: 3.6.8
* Flask version:
* Werkzeug version:

**Comment by shivamguys** [comment]
The reason is render_template_string assumes style="font-size:{{ font-size }}em"
font size is your key and it throws an error
![Screenshot from 2020-02-28 13-33-26](https://user-images.githubusercontent.com/25263989/75522054-0316a580-5a2f-11ea-8bec-34dd9bba3798.png)

I would advice you to use string format function where ever in your html uses key:value syntax like css.

Use render_template_string only where key:value dosen't exist in syntax, as a normal parser

**Comment by mitsuhiko** [comment]
Which version of Jinja is that?

**Comment by shivamguys** [comment]
Yes..
Jinja Version: 2.10.1
Werkzeug Version: 0.16.1

**Comment by davidism** [comment]
I'm having trouble reproducing this. I used your exact code and the mentioned versions, and got the expected traceback, not a reference loop or hang. Just in case, I also tested with the latest Werkzeug and Jinja versions and didn't see the issue either.

**Comment by davidism** [comment]
As @shivamguys showed, they didn't get an infinite loop either, they got the expected traceback that identified `'font' is undefined`.

**Comment by samuelchen** [comment]
> Which version of Jinja is that? The traceback must not have loops which to me would indicate that this is potentially a regression in Jinja.

@mitsuhiko   Jinja2      2.11.1

**Comment by samuelchen** [comment]
> The reason is render_template_string assumes style="font-size:{{ font-size }}em"
> font size is your key and it throws an error
> ![Screenshot from 2020-02-28 13-33-26](https://user-images.githubusercontent.com/25263989/75522054-0316a580-5a2f-11ea-8bec-34dd9bba3798.png)
> 
> I would advice you to use string format function where ever in your html uses key:value syntax like css.
> 
> Use render_template_string only where key:value dosen't exist in syntax, as a normal parser

Yes.. you are right. I was just trying to write an one file tool to render DB. It was supposed to throw exceptions out.

**Comment by samuelchen** [comment]
> As @shivamguys showed, they didn't get an infinite loop either, they got the expected traceback that identified `'font' is undefined`.

@davidism 
Win 10 64bit
Python 3.6.8  32bit
 Jinja2   2.11.1

hope it's helpful.

**Comment by samuelchen** [comment]
> I'm having trouble reproducing this. I used your exact code and the mentioned versions, and got the expected traceback, not a reference loop or hang. Just in case, I also tested with the latest Werkzeug and Jinja versions and didn't see the issue either.

I am not sure if my environment will impact it. But I tried run it directly in command line. It's the same.
